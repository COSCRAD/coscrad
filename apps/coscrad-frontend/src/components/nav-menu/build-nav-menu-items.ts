import { buildRoutes } from '../../app/build-routes';
import { ID_ROUTE_PARAM_KEY } from '../../app/routes';
import { ConfigurableContent } from '../../configurable-front-matter/data/configurable-content-schema';

const unlistedRoutes = ['*'];

export const buildNavMenuItems = (contentConfig: ConfigurableContent) => {
    const routes = buildRoutes(contentConfig);

    // TODO We should have a `shouldPublishInMenu` or `menus: string[]`  prop on CoscradRoute
    return routes
        .filter(({ path }) => !path.includes(`Resources/`))
        .filter(({ path }) => !unlistedRoutes.includes(path))
        .filter(({ path }) => !path.includes(ID_ROUTE_PARAM_KEY))
        .map(({ path, label }) => ({
            label,
            link: path,
        }));
};
