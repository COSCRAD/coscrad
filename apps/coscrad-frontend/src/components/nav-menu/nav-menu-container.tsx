import { CategorizableType } from '@coscrad/api-interfaces';
import { isNullOrUndefined } from '@coscrad/validation-constraints';
import { useContext } from 'react';
import { routes } from '../../app/routes/routes';
import { ConfigurableContentContext } from '../../configurable-front-matter/configurable-content-provider';
import { NavMenuPresenter } from './nav-menu-presenter';

export type NavItemInfo = {
    link: string;
    label: string;
};

export const NavMenuContainer = (): JSX.Element => {
    const { indexToDetailFlows, listenLive } = useContext(ConfigurableContentContext);
    // TODO: We should have a NavBar container and presenter

    // note this may be [] if we haven't included `notes`
    const dynamicLinks = indexToDetailFlows
        .filter(({ categorizableType }) => categorizableType === CategorizableType.note)
        .map(({ label, route }) => ({
            link: route || routes.notes.index,
            label: label || 'Notes',
        }))
        .concat([
            ...(isNullOrUndefined(listenLive)
                ? []
                : [
                      {
                          link: routes.listenLive,
                          label: listenLive.playingMessage,
                      },
                  ]),
        ]);

    // We may want an enum \ constants for our routes
    const navItemInfos: NavItemInfo[] = [
        {
            link: routes.home,
            label: 'Home',
        },
        {
            link: routes.about,
            label: 'About',
        },
        {
            link: routes.resources.info,
            label: 'Browse Resources',
        },
        {
            link: routes.tags.index,
            label: 'Tags',
        },
        {
            link: routes.treeOfKnowledge,
            label: 'Tree of Knowledge',
        },
        ...dynamicLinks,
        {
            link: routes.siteCredits,
            label: 'Credits',
        },
    ];

    return <NavMenuPresenter navItemInfos={navItemInfos} />;
};
