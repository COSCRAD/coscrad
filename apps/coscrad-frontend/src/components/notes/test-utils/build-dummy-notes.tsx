import {
    EdgeConnectionMemberRole,
    EdgeConnectionType,
    IMultilingualText,
    INoteViewModel,
    LanguageCode,
    MultilingualTextItemRole,
    ResourceCompositeIdentifier,
    ResourceType,
} from '@coscrad/api-interfaces';

const generalContext = {
    type: 'general',
};

export const buildMemberWithGeneralContext = (
    compositeIdentifier: ResourceCompositeIdentifier,
    role: EdgeConnectionMemberRole
) => ({
    context: generalContext,
    role,
    compositeIdentifier,
});

export const buildMultilingualTextFromEnglishOriginal = (text: string): IMultilingualText => ({
    items: [
        {
            text,
            role: MultilingualTextItemRole.original,
            languageCode: LanguageCode.English,
        },
    ],
});

const buildOneMemberForEveryResourceType = (role: EdgeConnectionMemberRole) =>
    Object.values(ResourceType).map((resourceType, index) =>
        buildMemberWithGeneralContext({ type: resourceType, id: index.toString() }, role)
    );

const selfEdgeConnections: INoteViewModel[] = buildOneMemberForEveryResourceType(
    EdgeConnectionMemberRole.self
).map((member, index) => ({
    connectionType: EdgeConnectionType.self,
    id: index.toString(),
    name: buildMultilingualTextFromEnglishOriginal(`name for self-note: ${index}`),
    note: buildMultilingualTextFromEnglishOriginal(`note for item ${index}`),
    connectedResources: [member],
}));

// TODO share formatters with back end
const formatCompositeIdentifier = ({ type, id }: ResourceCompositeIdentifier) => `${type}/${id}`;

export const buildDummyDualEdgeConnection = (
    fromCompositeIdentifier: ResourceCompositeIdentifier,
    toCompositeIdentifier: ResourceCompositeIdentifier,
    id: string
): INoteViewModel => ({
    connectionType: EdgeConnectionType.dual,
    id,
    name: buildMultilingualTextFromEnglishOriginal(
        `This is why ${formatCompositeIdentifier(
            fromCompositeIdentifier
        )} is connected to ${formatCompositeIdentifier(toCompositeIdentifier)}`
    ),
    note: buildMultilingualTextFromEnglishOriginal(
        `This is why ${formatCompositeIdentifier(
            fromCompositeIdentifier
        )} is connected to ${formatCompositeIdentifier(toCompositeIdentifier)}`
    ),
    connectedResources: [
        buildMemberWithGeneralContext(fromCompositeIdentifier, EdgeConnectionMemberRole.from),
        buildMemberWithGeneralContext(toCompositeIdentifier, EdgeConnectionMemberRole.to),
    ],
});

const dualEdgeConnections: INoteViewModel[] = [
    {
        id: '111',
        name: buildMultilingualTextFromEnglishOriginal('name for 111'),
        note: buildMultilingualTextFromEnglishOriginal(
            `This is why book 123 is related to media item 29`
        ),
        connectedResources: [
            buildMemberWithGeneralContext(
                {
                    type: ResourceType.digitalText,
                    id: '123789',
                },
                EdgeConnectionMemberRole.to
            ),
            buildMemberWithGeneralContext(
                {
                    type: ResourceType.mediaItem,
                    id: '29',
                },
                EdgeConnectionMemberRole.to
            ),
        ],
    },
    {
        id: '112',
        name: buildMultilingualTextFromEnglishOriginal('name for 112'),
        note: buildMultilingualTextFromEnglishOriginal(
            `This is why term 5 is related to audio item 8`
        ),
        connectedResources: [
            buildMemberWithGeneralContext(
                {
                    type: ResourceType.term,
                    id: '5',
                },
                EdgeConnectionMemberRole.to
            ),
            buildMemberWithGeneralContext(
                {
                    type: ResourceType.audioItem,
                    id: '8',
                },
                EdgeConnectionMemberRole.to
            ),
        ],
    },
    {
        id: '113',
        name: buildMultilingualTextFromEnglishOriginal('name for 113'),
        note: buildMultilingualTextFromEnglishOriginal(
            `This is why spatial feature 12 is related to photograph 293`
        ),
        connectedResources: [
            buildMemberWithGeneralContext(
                {
                    type: ResourceType.spatialFeature,
                    id: '12',
                },
                EdgeConnectionMemberRole.to
            ),
            buildMemberWithGeneralContext(
                {
                    type: ResourceType.photograph,
                    id: '293',
                },
                EdgeConnectionMemberRole.to
            ),
        ],
    },
].map((partial) => ({
    ...partial,
    connectionType: EdgeConnectionType.dual,
}));

export const buildDummyNotes = (desiredEdgeConnectionType?: EdgeConnectionType) => [
    ...(desiredEdgeConnectionType === EdgeConnectionType.self ? [] : dualEdgeConnections),
    ...(desiredEdgeConnectionType === EdgeConnectionType.dual ? [] : selfEdgeConnections),
];
