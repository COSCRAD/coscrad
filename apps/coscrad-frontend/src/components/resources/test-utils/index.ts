export * from './build-resource-detail-page-renderer-for-test';
export * from './render-resource-index-page-for-test';
