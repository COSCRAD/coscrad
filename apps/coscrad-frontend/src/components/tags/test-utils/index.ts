import { ITagViewModel } from '../../../../../../libs/api-interfaces/src';

export const buildDummyTags = (): ITagViewModel[] => [
    {
        label: 'birds',
        id: '201',
        // TODO Add members
        members: [],
    },
    {
        label: 'reptiles',
        id: '202',
        members: [],
    },
];
