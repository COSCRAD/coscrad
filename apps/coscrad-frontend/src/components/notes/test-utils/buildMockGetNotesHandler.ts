import { INoteViewModel } from '@coscrad/api-interfaces';
import { getConfig } from '../../../config';
import { buildMockSuccessfulGETHandler } from '../../../utils/test-utils/build-mock-successful-get-handler';
import { buildDummyNotes } from './build-dummy-notes';

export const buildMockGetNotesHandler = (notes?: INoteViewModel[]) =>
    buildMockSuccessfulGETHandler({
        endpoint: `${getConfig().apiUrl}/connections/notes`,
        response: notes || buildDummyNotes,
    });
