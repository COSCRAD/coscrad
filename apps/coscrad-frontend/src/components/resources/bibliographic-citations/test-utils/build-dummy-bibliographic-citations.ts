import {
    BibliographicCitationType,
    IBibliographicCitationViewModel,
    IBookBibliographicCitationData,
    ICourtCaseBibliographicCitationData,
    IJournalArticleBibliographicCitationData,
} from '@coscrad/api-interfaces';
import { BibliographicSubjectCreatorType } from '@coscrad/data-types';
import { buildMultilingualTextFromEnglishOriginal } from '../../../notes/test-utils';

// TODO Use BibliographicSubjectCreator enum here
const AUTHOR = BibliographicSubjectCreatorType.author;

const dummyBook: IBibliographicCitationViewModel<IBookBibliographicCitationData> = {
    name: buildMultilingualTextFromEnglishOriginal('A Day in the Life'),
    data: {
        type: BibliographicCitationType.book,
        title: 'A Day in the Life',
        creators: [
            {
                name: 'Alana Duvernay',
                type: AUTHOR,
            },
        ],
        abstract: 'This is the abstract, NOT a general note!',
        year: 1999,
        publisher: 'Atlantic Publishing',
        place: 'Kennebunk Port, Maine',
        url: 'https://atlanticpublishing.com',
        numberOfPages: 455,
        isbn: '978-1-895811-34-6',
    },
    id: '1',
    contributions: ['Rusty Smith'],
};

const dummyJournalArticle: IBibliographicCitationViewModel<IJournalArticleBibliographicCitationData> =
    {
        name: buildMultilingualTextFromEnglishOriginal(
            'Report on the Cariboo Chilcotin Justice inquiry'
        ),
        data: {
            type: BibliographicCitationType.journalArticle,
            title: 'Report on the Cariboo Chilcotin Justice inquiry',
            creators: [
                {
                    name: 'Sigurd Purcell',
                    type: AUTHOR,
                },
            ],
            abstract: 'An analysis of the Cariboo Chilcotin Justice inquiry.',
            issueDate: 'Spring 2013',
            publicationTitle: 'Journal of History',
            url: 'https://search.proquest.com/docview/1682229477/abstract/7836BCEA06014582PQ/1',
            issn: '00052949',
            doi: '10.14288/bcs.v0i19.784',
        },
        id: '23',
        contributions: ['Ron Doe'], // John's brother.
    };

const dummyCourtCase: IBibliographicCitationViewModel<ICourtCaseBibliographicCitationData> = {
    name: buildMultilingualTextFromEnglishOriginal("2002-07-08_Tsilhqot'inTitleCase"),
    data: {
        type: BibliographicCitationType.courtCase,
        caseName: "2002-07-08_Tsilhqot'inTitleCase",
        abstract:
            "William, June\n- Nemiah Valley\n- Qualifying as an expert in Tsilhqot'in written language",
        dateDecided: 'Recorded 2002-07-08',
        court: 'Supreme Court of British Columbia',
        url: 'https://www.myzoterolink.com/bogus-link-to-doc.php',
        pages: 'Pages 1-6',
    },
    id: '3',
    contributions: ['Jane Deer'],
};

export const buildDummyBibliographicCitations = (): IBibliographicCitationViewModel[] => [
    dummyBook,
    dummyJournalArticle,
    dummyCourtCase,
];
