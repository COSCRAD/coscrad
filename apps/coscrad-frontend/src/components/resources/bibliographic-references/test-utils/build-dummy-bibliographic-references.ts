import { BibliographicReferenceType } from '@coscrad/api-interfaces';
import { buildMultilingualTextFromEnglishOriginal } from '../../../notes/test-utils';

// TODO Use BibliographicSubjectCreator enum here
const AUTHOR = 'author';

export const buildDummyBibliographicReferences = () => [
    // Book
    {
        name: buildMultilingualTextFromEnglishOriginal('A Day in the Life'),
        data: {
            type: BibliographicReferenceType.book,
            title: 'A Day in the Life',
            creators: [
                {
                    name: 'Alana Duvernay',
                    type: AUTHOR,
                },
            ],
            abstract: 'This is the abstract, NOT a general note!',
            year: 1999,
            publisher: 'Atlantic Publishing',
            place: 'Kennebunk Port, Maine',
            url: 'https://atlanticpublishing.com',
            numberOfPages: 455,
            isbn: '978-1-895811-34-6',
        },
        id: '1',
    },
    // Journal Article
    {
        name: buildMultilingualTextFromEnglishOriginal(
            'Report on the Cariboo Chilcotin Justice inquiry'
        ),
        data: {
            type: BibliographicReferenceType.journalArticle,
            title: 'Report on the Cariboo Chilcotin Justice inquiry',
            creators: [
                {
                    name: 'Sigurd Purcell',
                    type: AUTHOR,
                },
            ],
            abstract: 'An analysis of the Cariboo Chilcotin Justice inquiry.',
            issueDate: 'Spring 2013',
            publicationTitle: 'Journal of History',
            url: 'https://search.proquest.com/docview/1682229477/abstract/7836BCEA06014582PQ/1',
            issn: '00052949',
            doi: '10.14288/bcs.v0i19.784',
        },
        published: true,
        id: '23',
    },
    // Court Case
    {
        name: buildMultilingualTextFromEnglishOriginal("2002-07-08_Tsilhqot'inTitleCase"),
        data: {
            type: BibliographicReferenceType.courtCase,
            caseName: "2002-07-08_Tsilhqot'inTitleCase",
            abstract:
                "William, June\n- Nemiah Valley\n- Qualifying as an expert in Tsilhqot'in written language",
            dateDecided: 'Recorded 2002-07-08',
            court: 'Supreme Court of British Columbia',
            url: 'https://www.myzoterolink.com/bogus-link-to-doc.php',
            pages: 'Pages 1-6',
        },
        published: true,
        id: '3',
    },
];
