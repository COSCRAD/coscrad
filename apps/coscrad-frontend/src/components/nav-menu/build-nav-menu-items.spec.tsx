import { isNonEmptyString } from '@coscrad/validation-constraints';
import { ConfigurableContent } from '../../configurable-front-matter/data/configurable-content-schema';
import { buildDummyConfig } from '../../utils/test-utils/build-dummy-content-config';
import { buildNavMenuItems } from './build-nav-menu-items';
import { NavItemInfo } from './nav-menu-container';

const liveRoute = 'Live';

const optionalPropertiesAndRoutesToExclude: [keyof ConfigurableContent, string][] = [
    ['listenLive', liveRoute],
];

const expectedOrder = [
    '/',
    'About',
    'Resources',
    'Notes',
    'Links',
    'TreeOfKnowledge',
    'Tags',
    liveRoute,
    'Credits',
];

const assertExpectedOrder = (info: NavItemInfo[]) => {
    const navInfoOrder = info.map(({ link }) => link);

    // Remove items from the primary list that are missing by virtue of being omitted in the config
    const expectedOrderWithMissingRemoved = expectedOrder.filter((link) =>
        navInfoOrder.includes(link)
    );

    // Now the two arrays should have the same members in the same order
    expect(navInfoOrder).toEqual(expectedOrderWithMissingRemoved);
};

describe(`dynamic navigation menu`, () => {
    describe(`when all properties are provided`, () => {
        const result = buildNavMenuItems(buildDummyConfig());
        it('should return the expected routes', () => {
            expect(result).toMatchSnapshot();
        });

        it('should have the correct order', () => {
            assertExpectedOrder(result);
        });

        it('should have the correct number of items', () => {
            expect(result.length).toBe(expectedOrder.length);
        });

        it('None of the labels should be empty', () => {
            const menuItemsWithEmptyLabels = result.filter(({ label }) => !isNonEmptyString(label));

            expect(menuItemsWithEmptyLabels).toEqual([]);
        });
    });

    optionalPropertiesAndRoutesToExclude.forEach(([key, route]) => {
        describe(`When the optional property: (${key}) is omitted from the content config`, () => {
            it(`should not include the corresponding route (${route})`, () => {
                const contentConfig = buildDummyConfig({
                    [key]: null,
                });

                const result = buildNavMenuItems(contentConfig);

                const isRouteIncluded = result.some(({ link }) => link === route);

                expect(isRouteIncluded).toBe(false);
            });
        });
    });

    describe(`When the Web of Knowledge is disabled`, () => {
        const routesToOmit = ['Notes', 'Tags', 'TreeOfKnowledge'];

        const contentConfig = buildDummyConfig({
            shouldEnableWebOfKnowledgeForResources: false,
        });

        const result = buildNavMenuItems(contentConfig);

        it(`should not include any of: ${routesToOmit.join(',')}`, () => {
            const extraRoutes = result.filter(({ link }) => routesToOmit.includes(link));

            expect(extraRoutes).toEqual([]);
        });

        it(`should have the expected order`, () => {
            assertExpectedOrder(result);
        });
    });
});
