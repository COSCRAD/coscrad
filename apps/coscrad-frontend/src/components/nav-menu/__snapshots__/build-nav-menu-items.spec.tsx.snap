// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dynamic navigation menu when all properties are provided should return the expected routes 1`] = `
[
  {
    "label": "Home",
    "link": "/",
  },
  {
    "label": "About",
    "link": "About",
  },
  {
    "label": "Available Resources",
    "link": "Resources",
  },
  {
    "label": "Notes",
    "link": "Notes",
  },
  {
    "label": "Tree of Knowledge",
    "link": "TreeOfKnowledge",
  },
  {
    "label": "Tags",
    "link": "Tags",
  },
  {
    "label": "89.5",
    "link": "Live",
  },
  {
    "label": "Credits",
    "link": "Credits",
  },
]
`;
